core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/evaluationsteps/LeftJoinQueryEvaluationStep.java

```diff
@@ -61,18 +55,60 @@ public CloseableIteration<BindingSet, QueryEvaluationException> evaluate(Binding
 		} else {
 			condition = null;
 		}
-		return new LeftJoinQueryEvaluationStep(right, condition, left, leftJoin, optionalVarCollector);
+		return new LeftJoinQueryEvaluationStep(right, condition, left, leftJoin, optionalVarCollector.getVarNames());
 	}
 
+	@Deprecated(forRemoval = true, since = "4.1.0")
 	public LeftJoinQueryEvaluationStep(QueryEvaluationStep right, QueryValueEvaluationStep condition,
-			QueryEvaluationStep left, LeftJoin leftJoin, VarNameCollector optionalVarCollector) {
+			QueryEvaluationStep left, LeftJoin leftJoin,
+			org.eclipse.rdf4j.query.algebra.helpers.VarNameCollector optionalVarCollector) {
 		this.right = right;
 		this.condition = condition;
 		this.left = left;
 		this.leftJoin = leftJoin;
 		// This is used to determine if the left join is well designed.
-		this.optionalVars = optionalVarCollector.getVarNames();
-		this.optionalVars.removeAll(leftJoin.getLeftArg().getBindingNames());
+
+		Set<String> leftBindingNames = leftJoin.getLeftArg().getBindingNames();
+		Set<String> optionalVars = optionalVarCollector.getVarNames();
+
+		boolean optionalVarsContainLeftBindingName = false;
+		for (String leftBindingName : leftBindingNames) {
+			if (!optionalVarsContainLeftBindingName && optionalVars.contains(leftBindingName)) {
+				optionalVars = new HashSet<>(optionalVars);
+				optionalVarsContainLeftBindingName = true;
+			}
+			if (optionalVarsContainLeftBindingName) {
+				optionalVars.remove(leftBindingName);
+			}
+		}
+
+		this.optionalVars = optionalVars;
+
+	}
+
+	public LeftJoinQueryEvaluationStep(QueryEvaluationStep right, QueryValueEvaluationStep condition,
+			QueryEvaluationStep left, LeftJoin leftJoin, Set<String> optionalVars) {
+		this.right = right;
+		this.condition = condition;
+		this.left = left;
+		this.leftJoin = leftJoin;
+		// This is used to determine if the left join is well designed.
+
+		Set<String> leftBindingNames = leftJoin.getLeftArg().getBindingNames();
+
+		boolean optionalVarsContainLeftBindingName = false;
+		for (String leftBindingName : leftBindingNames) {
+			if (!optionalVarsContainLeftBindingName && optionalVars.contains(leftBindingName)) {
+				optionalVars = new HashSet<>(optionalVars);
+				optionalVarsContainLeftBindingName = true;
+			}
+			if (optionalVarsContainLeftBindingName) {
+				optionalVars.remove(leftBindingName);
+			}
+		}
+
+		this.optionalVars = optionalVars;
+
```

Comment:
This allows the var name collector to cache the set it returns instead of creating a new one each time.